src/couplings.jl

```diff
@@ -0,0 +1,244 @@
+
+# Examples
+```julia-repl
+julia> m = PartitionMask(3, [1], [2]) # <= assumes input-length 3
+PartitionMask{SparseArrays.SparseMatrixCSC{Float64,Int64}}(
+  [1, 1]  =  1.0,
+  [2, 1]  =  1.0,
+  [3, 1]  =  1.0)
+
+julia> # Partition into 3 parts; the last part is inferred to be indices `[3, ]` from
+       # the fact that `[1]` and `[2]` does not make up all indices in `1:3`.
+       x1, x2, x3 = partition(m, [1., 2., 3.])
+([1.0], [2.0], [3.0])
+
+julia> # Recombines the partitions into a vector
+       combine(m, x1, x2, x3)
+3-element Array{Float64,1}:
+ 1.0
+ 2.0
+ 3.0
+```
+
+"""
+struct PartitionMask{A}
+    A_1::A
+    A_2::A
+    A_3::A
+
+    # Only make it possible to construct using matrices
+    PartitionMask(A_1::A, A_2::A, A_3::A) where {A <: AbstractMatrix{<:Real}} = new{A}(A_1, A_2, A_3)
+end
+
+function PartitionMask(
+    n::Int,
+    indices_1::AbstractVector{Int},
+    indices_2::AbstractVector{Int},
+    indices_3::AbstractVector{Int}
+)
+    A_1 = spzeros(n, length(indices_1));
+    A_2 = spzeros(n, length(indices_2));
+    A_3 = spzeros(n, length(indices_3));
+
+    for (i, idx) in enumerate(indices_1)
+        A_1[idx, i] = 1.0
+    end
+
+    for (i, idx) in enumerate(indices_2)
+        A_2[idx, i] = 1.0
+    end
+
+    for (i, idx) in enumerate(indices_3)
+        A_3[idx, i] = 1.0
+    end
+
+    return PartitionMask(A_1, A_2, A_3)
+end
+
+PartitionMask(
+    n::Int,
+    indices_1::AbstractVector{Int},
+    indices_2::AbstractVector{Int}
+) = PartitionMask(n, indices_1, indices_2, nothing)
+
+PartitionMask(
+    n::Int,
+    indices_1::AbstractVector{Int},
+    indices_2::AbstractVector{Int},
+    indices_3::Nothing
+) = PartitionMask(n, indices_1, indices_2, [i for i in 1:n if i ∉ (indices_1 ∪ indices_2)])
+
+PartitionMask(
+    n::Int,
+    indices_1::AbstractVector{Int},
+    indices_2::Nothing,
+    indices_3::AbstractVector{Int}
+) = PartitionMask(n, indices_1, [i for i in 1:n if i ∉ (indices_1 ∪ indices_3)], indices_3)
+
+"""
+    PartitionMask(n::Int, indices)
+
+Assumes you want to _split_ the vector, where `indices` refer to the 
+parts of the vector you want to apply the bijector to.
+"""
+function PartitionMask(n::Int, indices)
+    indices_2 = [i for i in 1:n if i ∉ indices]
+
+    # sparse arrays <3
+    A_1 = spzeros(n, length(indices));
+    A_2 = spzeros(n, length(indices_2));
+
+    # Like doing:
+    #    A[1, 1] = 1.0
+    #    A[3, 2] = 1.0
+    for (i, idx) in enumerate(indices)
+        A_1[idx, i] = 1.0
+    end
+
+    for (i, idx) in enumerate(indices_2)
+        A_2[idx, i] = 1.0
+    end
+
+    return PartitionMask(A_1, A_2, spzeros(n, 0))
+end
+PartitionMask(x::AbstractVector, indices) = PartitionMask(length(x), indices)
+
+"""
+    combine(m::PartitionMask, x_1, x_2, x_3)
+
+Combines `x_1`, `x_2`, and `x_3` into a single vector.
+"""
+@inline combine(m::PartitionMask, x_1, x_2, x_3) = m.A_1 * x_1 .+ m.A_2 * x_2 .+ m.A_3 * x_3
```

Comment:
Nice use of sparse matrices and transposes but I doubt this will be faster or more Tracker and GPU friendly than simply `collect(sparsevec(inds1, x1, n)) + collect(sparsevec(inds2, x2, n)) + ...`. Some of these functions may need to be defined for `TrackedArray` or `GPUArray` but nothing should stop that in principle. Also do the indices need to be all over the place, or can we assume contiguity of the partitions?

Reply:
It should be more GPU-friendly though, or at least make it easier to support GPUs, since you can instantiate `PartitionMask` with the corresponding GPU arrays. E.g. `PartitionMask(cu(A_1), cu(A_2), cu(A_3))`, right?

Reply:
> Also do the indices need to be all over the place, or can we assume contiguity of the partitions?

I introduce `Permute <: Bijector` in a different PR, so we _could_ assume contiguity of partitions with some additional work to make things easier on the user. But is this worth it? I guess with my current implementation it won't make much difference, but do you have different approach that makes this more efficient (assuming contiguity of partitions)?

Also, I'll have a GPU available to work on soon, at which point I'll start making everything compatible with GPU.

Reply:
Sorry I missed your comments. You can try both and see if the simple version works on the GPU.

Reply:
```julia
using Test

using Bijectors
using Bijectors: CouplingLayer, Shift, Scale, PartitionMask, partition, combine

### On GPU
using CuArrays, SparseArrays
x = Float32.(randn(100_000))
x_gpu = cu(x)

mask = PartitionMask(length(x), collect(1:100), collect(500:1000))
# Convert directly into `CuSparseMatrixCSC` instead of doing `sparse ∘ cu`
# due to the fact that `cu` will first load the full non-sparse matrix into
# memory before converting into sparse using `sparse`.
mask_gpu = PartitionMask(
    CUSPARSE.CuSparseMatrixCSC.(
        convert.(SparseMatrixCSC{Float32, Int}, (mask.A_1, mask.A_2, mask.A_3))
    )...
);

x1, x2, x3 = partition(mask_gpu, x_gpu)
y = combine(mask_gpu, x1, x2, x3)
```
and then
```julia
using BenchmarkTools
@btime begin
    x1, x2, x3 = partition($mask_gpu, $x_gpu)
    y = combine($mask_gpu, x1, x2, x3)
end 
# Memory usage is wrong since not accounting for GPU memory
# => 1.488 ms (203 allocations: 7.22 KiB)

@btime begin
    x1, x2, x3 = partition($mask, $x)
    y = combine($mask, x1, x2, x3)
end
# => 573.984 μs (15 allocations: 3.82 MiB)
```
Works without any issues. Though I ran into an issue where if I was using `Float64` for the sparse matrix, `transpose(mask_gpu.A_1) * x_gpu` would silently do _nothing_. This is of course an issue with `CUSPARSE.jl` rather than with `Bijectors.jl`. 

With that being said, the above is about 3X slower on the GPU than on CPU. Not sure whether or not this is expected?